lib/charms/observability_libs/v0/juju_topology.py

```diff
@@ -0,0 +1,230 @@
+            model_uuid=charm.model.uuid,
+            application=charm.model.app.name,
+            unit=charm.model.unit.name,
+            charm_name=charm.meta.name,
+        )
+
+    @classmethod
+    def from_dict(cls, data: dict):
+        """Factory method for creating `JujuTopology` children from a dictionary.
+
+        Args:
+            data: a dictionary with five keys providing topology information. The keys are
+                - "model"
+                - "model_uuid"
+                - "application"
+                - "unit"
+                - "charm_name"
+                `unit` and `charm_name` may be empty, but will result in more limited
+                labels. However, this allows us to support payload-only charms.
+
+        Returns:
+            a `JujuTopology` object.
+        """
+        return cls(
+            model=data["model"],
+            model_uuid=data["model_uuid"],
+            application=data["application"],
+            unit=data.get("unit", ""),
+            charm_name=data.get("charm_name", ""),
+        )
+
+    def as_dict(
+        self,
+        *,
+        remapped_keys: Dict[str, str] = {},
+        excluded_keys: List[str] = [],
+        uuid_length: Optional[int] = None,
+    ) -> OrderedDict:
+        """Format the topology information into an ordered dict.
+
+        Keeping the dictionary ordered is important to be able to
+        compare dicts without having to resort to deep comparisons.
+
+        Args:
+            remapped_keys: A dictionary mapping old key names to new key names,
+                which will be substituted when invoked.
+            excluded_keys: A list of key names to exclude from the returned dict.
+            uuid_length: The length to crop the UUID to.
+        """
+        model_uuid = self.model_uuid[: uuid_length - 1] if uuid_length else self.model_uuid
+        ret = OrderedDict(
+            [
+                ("model", self.model),
+                ("model_uuid", model_uuid),
+                ("application", self.application),
+                ("unit", self.unit),
+                ("charm_name", self.charm_name),
+            ]
+        )
+
+        for exclusion in excluded_keys:
+            if exclusion in ret:
+                ret.pop(exclusion)
+
+        if remapped_keys:
+            ret = OrderedDict(
+                (remapped_keys.get(k), v) if remapped_keys.get(k) else (k, v) for k, v in ret.items()  # type: ignore
+            )
+
+        return ret
+
+    def as_identifier(self) -> str:
+        """Format the topology information into a terse string.
+
+        This crops the model UUID, making it not suitable for comparisons against
+        anything but other identifiers.
+        """
+        values = self.as_dict(
+            excluded_keys=["unit", "charm_name"],
+            uuid_length=8,
+        ).values()
+
+        return "_".join([str(val) for val in values]).replace("/", "_")
+
+    def as_relabelled_dict(self):
```

Comment:
I would drop this function in favor of using `as_dict` with a new additional arg `key_mutator=lambda k: "juju_" + k`.

Reply:
If the naming is bad, I'm all for renaming it, but I do agree with Ryan that we should be opinionated. Any flexibility should be out of necessity, not because it is nice to have.

Reply:
In that case... `ObservabilityTopology`? Or just drop "Juju" and call this "Topology" (as it comes from observability_libs).

Reply:
Mentioned on Meet, but we want this to go into OF if we can. Even if it ultimately gets exposed as `charm.topology = FooBarBazTopology.from_charm(self)`, we may as well wait to agonize over an appropriate name until those conversations start to converge.

Reply:
I am also in favour of dropping this method and refactoring related code to always having `juju_` prefix.

Reply:
That is not an option, unfortunately. Adding a `key_mutator` function further limits the usability of having this library in the first place (ie. not having to think too much about how it *should* be)